### ISSUE TITLE
Add schema/Schematron checks to ensure that extended content always has a web accessible schema definition

### ISSUE THREAD

**Issue by braeckel** [wmo-im/iwxxm#29]
Add checks to ensure that all extended content blocks have a web-accessible XML schema.  Without this check it is possible to leave off the schemaLocation and the validator will not attempt to validate extended content without a schema.  This is a safety issue in that the extended content essentially cannot be validated and has no description (including human-readable description) for consumers to use.  There are two approaches:

1) Use processContents="strict" on all extension elements.  This may have the desired behavior but more research required.  This may require making use of the xs:any element instead of the current anyType

2) Add a Schema or Schematron check to ensure that the top-level element under each extension element has an xsi:schemaLocation for that top-level element’s namespace.  For example, iwxxm-us elements should have an xsi:schemaLocation on each element that indicates the location for the iwxxm-us namespace.  If a schemaLocation is set then validation will ensure that it is web-accessible and correct.  This would not guarantee that all contents are validated, just the top element, but if this was extended for all namespaces under the extension element might get tricky to implement (increasing the need for a namespace "whitelist", for example).

**Comment by blchoy** [comment]
I am not an expert in validators but mine Oxygen XML Editor always complain if a name space is not defined and if defined its schema location is not accessible.

In reality, while most of us will read IWXXM reports close to their generation time, there are always times when we need to read these reports much longer afterwards (e.g. investigation).  Therefore we need a mechanism to preserve the schemas referenced by the extensions.  Whether this mechanism is owned by WMO/ICAO is another story.

**Comment by braeckel** [comment]
The validator will always ensure that namespace prefixes are defined but if processContents="strict" is used it will ensure that the schemaLocations are also valid.  

For an example take the following snippet and put it at the bottom of any IWXXM message.  It validates despite having an unresolvable (made up) namespace.

```XML
<iwxxm:extension xmlns:foo="http://i-do-not-exist.invalid">
  <foo:content>It works!</foo:content>
</iwxxm:extension>
```

However I just did a test with SIGMET and changed the existing extension element definitions from:

```XML
<element name="extension" minOccurs="0" maxOccurs="unbounded" type="anyType">
  <annotation>
     <documentation>Extension block for optional and/or additional parameters for element SIGMET</documentation>
    </annotation>
</element>
```

to the following custom ExtensionType:

```XML
<element name="extension" minOccurs="0" maxOccurs="unbounded" type="iwxxm:ExtensionType">
    <annotation>
      <documentation>Extension block for optional and/or additional parameters for element SIGMET</documentation>
    </annotation>
</element>

<complexType name="ExtensionType">
  <sequence>
    <any processContents="strict"/>
  </sequence>
</complexType>
```

And this updated version properly causes a validation failure because the schema could not be located.  I think the processContents="strict" on <xs:any> is a great solution for this issue.

**Comment by blchoy** [comment]
Noted and agreed.  However, some documentation (https://www.w3schools.com/xml/el_any.asp) said processContents="strict" should be the default.  :)

**Comment by braeckel** [comment]
I also saw mentions of that being the default behavior, but in testing with
Oxygen and Crux (both are Xerces Java) that doesn't seem to be the case.

**Comment by braeckel** [comment]
The most sensible approach would seem to be to:
 - add an XSLT rule in the pre/post processing to add the following to common.xsd:

```
<complexType name="ExtensionType">
  <sequence>
    <any processContents="strict"/>
  </sequence>
</complexType>
```

 - change the existing XSLT pre/post processing logic to use this ExtensionType for each extension element currently being generated
 - modify the XSLT pre/post processing to ensure that every IWXXM .xsd file imports common.xsd

**Comment by blchoy** [comment]
The EA post-processing XSLT has been modified and committed to the SVN.

**Comment by blchoy** [comment]
I am closing this as I see there is no further comment from Aaron on my implementation of the EA post-processing XSLT scripts.

**Comment by braeckel** [comment]
When the strict behavior is enabled without a resolvable schema location Xerces returns an error that says: "Failed to read schema document 'http://host.org/schema/iwxxm-ext/1.0/iwxxm-ext.xsd', because 1) could not find the document; 2) the document could not be read; 3) the root element of the document is not <xsd:schema>.'

**Comment by braeckel** [comment]
In TT-AvXML/7 it was agreed that comments on this strict behavior would be sought and perhaps loosened in 3.0RCfinal

**Comment by blchoy** [comment]
No active complaints so far but we still need to do the research on how this works on AIXM according to MIE/4.  This won't probably be resolved in 3.0.

**Comment by blchoy** [comment]
At both WG-MIE/5 and TT-AvXML-8 there were general consensus in this arrangement.  This issue is hereby closed.
